# Guided Project: Exploring eBay Car Sales Data

## Introduction to Data

In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website. We've made a few modifications from the original dataset:

* We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
* We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

The aim of this project is to clean the data and analyze the included used car listings. You'll also become familiar with some of the unique benefits jupyter notebook provides for pandas. 

### Instructions

1. Import the pandas and NumPy libraries
2. Read the `autos.csv` CSV file into pandas, and assign it to the variable name `autos`.
    * Try without specifying any encoding (which will default to `UTF-8`)
    * If you get an encoding error, try the next two most popular encodings (`Latin-1` and `Windows-1252`) until you are able to read the file without error.
3. Create a new cell with just the variable `autos` and run this cell.
    * A neat feature of jupyter notebook is its ability to render the first few and last few values of any pandas object.
4. Use the `DataFrame.info()` and `DataFrame.head()` methods to print information about the `autos` dataframe, as well as the first few rows.
    * Write a markdown cell briefly describing your observations.

In [1]:
# Import pandas and NumPy libraries
import pandas as pd
import numpy as np

# Read autos.csv into pandas and assign it to autos
autos = pd.read_csv('autos.csv', encoding = "Latin-1") # Had to use "Latin-1" encoding because default UTF-8 had error

In [2]:
# Creat new cell with just autos variables
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
# Get info of autos variable and use head variable
autos.info()

autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

There are a lot of object datatypes and the ones that are int64 seem like they need to be that datatype

## Cleaning Column Names

From the work we did in the last screen, we can make the following observations:

* The dataset contains 20 columns, most of which are strings.
* Some columns have null values, but none have more than ~20% null values.
* The column names use [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case), which means we can't just replace spaces with underscores.

### Instructions

1. Use the `DataFrame.columns` attribute to print an array of the existing column names.
2. Copy that array and make the following edits to columns names:
    * `yearOfRegistration` to `registration_year`
    * `monthOfRegistration` to `registration_month`
    * `notRepairedDamage` to `unrepaired_damage`
    * `dateCreated` to `ad_created`
    * The rest of the column names from camelcase to snakecase.
3. Assign the modified column names back to the `DataFrame.columns` attribute.
4. Use `DataFrame.head()` to look at the current state of the `autos` dataframe.
5. Write a markdown cell explaining the changes you made and why.

In [4]:
# Use columns attribe to print array of existing column name
print(autos.columns)

# Change column names
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 
                 'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month',
                 'fuel_type','brand', 'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
                 'last_seen']

# Use dataframe.head() to look at current state of autos dataframe
autos.head()

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

I made these changes to make column names into snakecase which puts an underscore in the name and makes it easier to read and understand the column names.

## Initial Exploration and Cleaning

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:

* Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
* Examples of numeric data stored as text which can be cleaned and converted.

The following methods are helpful for exploring the data:

* `DataFrame.describe()` (with `include='all'` to get both categorical and numeric columns)
* `Series.value_counts()` and `eries.head()` if any columns need a closer look.

### Instructions

1. Use `DataFrame.describe()` to look at descriptive statistics for all columns.
2. Write a markdown cell noting:
    * Any columns that have mostly one value that are candidates to be dropped
    * Any columns that need more investigation.
    * Any examples of numeric data stored as text that needs to be cleaned.
3. If you need to investigate any columns more, do so and write up any additional things you found.
4. You likely found that the `price` and `odometer` columns are numeric values stored as text. For each column:
    * Remove any non-numeric characters.
    * Convert the column to a numeric dtype.
    * Use `DataFrame.rename()` to rename the column to `odometer_km`.

In [5]:
# Use DataFrame.describe() to look at descriptive statistics
autos.describe(include='all')

print(autos["price"].value_counts().head()) #stored as text
print(autos["odometer"].value_counts().head()) # stored as text

# Remove non_numeric characters from price
autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "").astype(int)
## Check to make sure it worked
print(autos["price"])

# Remove non_numeric characters from odometer and change name
autos["odometer"] = autos["odometer"].str.replace(",", "").str.replace("km", "").astype(int)
autos.rename({"odometer":"odometer_km"}, axis = 1, inplace=True)
## Check to make sure it worked
print(autos["odometer_km"])


$0        1421
$500       781
$1,500     734
$2,500     643
$1,200     639
Name: price, dtype: int64
150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
Name: odometer, dtype: int64
0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64
0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer_km, Length: 50000, dtype: int64


## Exploring the Odometer and Price Columns

From the last screen, we learned that there are a number of text columns where almost all of the values are the same (`seller` and `offer_type`). We also converted the `price` and `odometer` columns to numeric types and renamed `odometer` to `odometer_km`.

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the `odometer_km` and `price` columns. Here's the steps we'll take:

* Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.
* We'll use:
    * `Series.unique().shape` to see how many unique values
    * `Series.describe()` to view min/max/median/mean etc
    * `Series.value_counts()`, with some variations:
        * chained to `.head()` if there are lots of values.
        * Because `Series.value_counts()` returns a series, we can use `Series.sort_index()` with `ascending=` `True` or `False` to view the highest and lowest values with their counts (can also chain to `head()` here).
    * When removing outliers, we can do `df[(df["col"] >= x ) & (df["col"] <= y )]`, but it's more readable to use `df[df["col"].between(x,y)]`
    
### Instructions
1. For each of `odometer_km` and `price` columns:
    * Use the techniques above to explore the data
    * If you find there are outliers, remove them and write a markdown paragraph explaining your decision.
    * After you have removed the outliers, make some observations about the remaining values.

In [6]:
# How many unique values in odometer_km column
unique_odometer_km = autos["odometer_km"].unique().shape
print(unique_odometer_km)

(13,)


In [7]:
# Describing odometer_km column
describe_odometer_km = autos["odometer_km"].describe()
print(describe_odometer_km)

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [8]:
# Determining if there are outliers in odometer_km column using top 10 highest speeds
value_counts_odometer_km = autos["odometer_km"].value_counts().sort_index(ascending=False).head(10)
print(value_counts_odometer_km)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
Name: odometer_km, dtype: int64


In [9]:
# Determining if there are outliers in odometer_km column using lowest 10 speeds
value_counts_odometer_km = autos["odometer_km"].value_counts().sort_index(ascending=True).head(10)
print(value_counts_odometer_km)

5000      967
10000     264
20000     784
30000     789
40000     819
50000    1027
60000    1164
70000    1230
80000    1436
90000    1757
Name: odometer_km, dtype: int64


It does not seem like there are outliers in odometer_km column

In [10]:
# Analyze price columns
unique_price = autos["price"].unique().shape
print(unique_price)

(2357,)


In [11]:
# Describe price column
describe_price = autos["price"].describe()
print(describe_price)

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [12]:
# Determine if there are outliers using top 10 highest prices
values_counts_price = autos["price"].value_counts().sort_index(ascending=False).head(10)
print(values_counts_price)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64


In [13]:
# Determine if there are outliers using lowest 20 prices
values_counts_price = autos["price"].value_counts().sort_index(ascending=True).head(20)
print(values_counts_price)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64


Defintely seems to outliers. There is a large jump after 130K and there seems to be a lot of zeros which may be an error as well. Moreover, eBay is often a bidding website, so bidding can start at as low as 1 dollar, which could explain the count of 156 for 1 dollar - and suggest the other low prices above 0 are in fact real.

I think we can safely omit cars with prices above 350k from our analysis: as the counts are consistently low this will only mean removing 14 lines of data, whilst removing these huge price numbers that could skew the data.

In [14]:
# Removing outliers from price column
autos = autos[autos["price"].between(1,350000)]

# Decribes new price column
describe_price = autos["price"].describe()
print(describe_price)

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64


## Exploring the date columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like `Series.describe()` to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [15]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

You'll notice that the first 10 characters represent the day (e.g. `2016-03-12`). To understand the date range, we can extract just the date values, use `Series.value_counts()` to generate a distribution, and then sort by the index.

To select the first 10 characters in each column, we can use `Series.str[:10]`:

In [16]:
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object


### Instructions

1. Use the workflow we just described to calculate the distribution of values in the `date_crawled`, `ad_created`, and `last_seen` columns (all string columns) as percentages.
    * To include missing values in the distribution and to use percentages instead of counts, chain the `Series.value_counts(normalize=True, dropna=False)` method.
    * To rank by date in ascending order (earliest to latest), chain the `Series.sort_index()` method.
    * Write a markdown cell after each column exploration to explain your observations.
2. Use `Series.describe()` to understand the distribution of `registration_year`.
    * Write a markdown cell explaining your observations.

In [17]:
# Date crawled column
(autos["date_crawled"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The crawled dates are every day between (and including) 5th March and 7th April 2016 - with around 3% of cars crawled on most of the individual dates.

In [18]:
# Ad created column
(autos["ad_created"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

There are 76 different dates that ads were created on - ranging from just under nine months before the earliest `date_crawled` date (11 June 2015 vs 5 March 2016), to the date of the latest `date_crawled` date (7 April 2016). Based on the percentages of the five earliest and five latest dates we can see above, it looks like more of the ads created are close to the `date_crawled` period.

In [19]:
# Last seen columns
(autos["last_seen"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

As a reminder this column is "When the crawler saw this ad last online" - so this must mean the `last_seen` dates will be the date that either (a) the car is sold, (b) the data for the car is crawled for the last time, or (c) the car ad is taken down (though I'm not sure if that means there would be no data at all). (b) likely accounts for the latest dates being in the magnitude of 12-22%, versus 1-3% for most other dates i.e. the crawling ended, instead of much more cars being sold or ads being taken down on those dates.

In [20]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We haven't made it to the apparent max year of 9999 yet (just another 7976 years to go), and cars definitely did not yet exist in the apparent min year of 1000, so there is clearly some incorrect data here:

* Given all data was crawled in 2016, it also doesn't make sense to have any `registration_year` higher than 2016.
* A quick look on Google tells me "On January 29, 1886, Carl Benz applied for a patent for his “vehicle powered by a gas engine.” The patent – number 37435 – may be regarded as the birth certificate of the automobile", so we can likely safely remove any data with a `registration_year` below 1886.

## Dealing with Incorrect Registration Year Data

One thing that stands out from the exploration we did in the last screen is that the registration_year column contains some odd values:

* The minimum value is 1000, before cars were invented
* The maximum value is 9999, many years into the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

### Instructions

1. Decide which the highest and lowest acceptable values are for the registration_year column.
    * Write a markdown cell explaining your decision and why
2. Remove the values outside those upper and lower bounds and calculate the distribution of the remaining values using Series.value_counts(normalize=True).
    * Write a markdown cell explaining your decision and why

In [21]:
# Remove upper and lower bounds for registration_year
autos = autos[autos["registration_year"].between(1990,2016)]

In [22]:
autos["registration_year"].value_counts(normalize=True)

2000    0.069540
2005    0.064692
1999    0.063833
2004    0.059558
2003    0.059470
2006    0.058831
2001    0.058082
2002    0.054777
1998    0.052067
2007    0.050172
2008    0.048806
2009    0.045941
1997    0.042989
2011    0.035762
2010    0.035012
1996    0.030253
2012    0.028865
1995    0.027036
2016    0.026882
2013    0.017693
2014    0.014609
1994    0.013860
1993    0.009365
2015    0.008637
1992    0.008153
1990    0.007646
1991    0.007470
Name: registration_year, dtype: float64

## Exploring Price by Brand

One of the analysis techniques we learned in this course is aggregation. When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the `brand` column.

If you recall in an earlier lesson, we explored how to use loops to perform aggregation. Here's what the process looks like:

* Identify the unique values we want to aggregate by
* Create an empty dictionary to store our aggregate data
* Loop over the unique values, and for each: - Subset the dataframe by the unique values - Calculate the mean of whichever column we're interested in - Assign the val/mean to the dict as k/v.

### Instructions
1. Explore the unique values in the `brand` column, and decide on which brands you want to aggregate by. 
    * You might want to select the top 20, or you might want to select those that have over a certain percentage of the total values (e.g. > 5%).
    * Remember that `Series.value_counts()` produces a series with index labels, so you can use `Series.index` attribute to access the labels, should you wish.
2. Write a short paragraph describing the brand data, and explaining which brands you've chosen to aggregate on.
3. Create an empty dictionary to hold your aggregate data.
    * Loop over your selected brands, and assign the mean price to the dictionary, with the brand name as the key.
    * Print your dictionary of aggregate data, and write a paragraph analyzing the results.

In [23]:
# Explore brand column
autos["brand"].value_counts()

volkswagen        9572
bmw               5046
opel              4938
mercedes_benz     4253
audi              3993
ford              3176
renault           2185
peugeot           1392
fiat              1164
seat               852
skoda              761
nissan             704
mazda              704
smart              661
citroen            637
toyota             583
hyundai            468
volvo              406
mini               405
mitsubishi         382
honda              356
kia                330
sonstige_autos     320
alfa_romeo         290
suzuki             275
porsche            249
chevrolet          234
chrysler           159
dacia              123
daihatsu           116
jeep               102
subaru              96
land_rover          94
saab                73
jaguar              70
daewoo              70
rover               62
lancia              48
lada                22
trabant             13
Name: brand, dtype: int64

In [24]:
# Percentages
autos["brand"].value_counts(normalize=True)

volkswagen        0.210911
bmw               0.111185
opel              0.108805
mercedes_benz     0.093711
audi              0.087983
ford              0.069981
renault           0.048145
peugeot           0.030672
fiat              0.025648
seat              0.018773
skoda             0.016768
nissan            0.015512
mazda             0.015512
smart             0.014565
citroen           0.014036
toyota            0.012846
hyundai           0.010312
volvo             0.008946
mini              0.008924
mitsubishi        0.008417
honda             0.007844
kia               0.007271
sonstige_autos    0.007051
alfa_romeo        0.006390
suzuki            0.006059
porsche           0.005487
chevrolet         0.005156
chrysler          0.003503
dacia             0.002710
daihatsu          0.002556
jeep              0.002247
subaru            0.002115
land_rover        0.002071
saab              0.001608
jaguar            0.001542
daewoo            0.001542
rover             0.001366
l

Volkswagen makes up a huge ~ 21%, at least 10% more than the next six most common brands: BMW (~ 11%), Opel (~ 10%), Mercedes (~ 9%), Audi (~ 9%), Ford (~ 7%) and Renault (~ 5%). After Renault, we have Peugeot at ~ 3%, with the numbers gradually decreasing to as low as ~ 0.03% for Trabant. Just going to just use the top 7 brands. 

In [25]:
# Top brands
top_brands_count = (autos["brand"].value_counts(normalize = True).head(7))
top_brands = top_brands_count.index

# Create empty dictionary
mean_brand_price = {}

for b in top_brands:
    selected_rows = autos[autos["brand"] == b]
    mean = selected_rows["price"].mean()
    brand_dict[b] = mean

print(mean_brand_price)

{'volkswagen': 5398.856038445466, 'bmw': 8361.343440348792, 'opel': 2953.5149858242203, 'mercedes_benz': 8582.339054784858, 'audi': 9420.485098923116, 'ford': 3448.153967254408, 'renault': 2420.4929061784896}


## Storing Aggregate Data in a DataFrame

In the last screen, we aggregated across brands to understand mean price. We observed that in the top 6 brands, there's a distinct price gap.

* Audi, BMW and Mercedes Benz are more expensive
* Ford and Opel are less expensive
* Volkswagen is in between

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

* it's difficult to compare more than two aggregate series objects if we want to extend to more columns
* we can't compare more than a few rows from each series object
* we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Instead, we can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly. To do this, we'll need to learn two pandas methods:

* [pandas series constructor](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html)
* [pandas dataframe constructor](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html)

Here's an example of the series constructor that uses the `brand_mean_prices` dictionary:

`
bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)
volkswagen       5402
bmw              8332
opel             2975
mercedes_benz    8628
audi             9336
ford             3749
dtype: int64
`

The keys in the dictionary became the index in the series object. We can then create a single-column dataframe from this series object. We need to use the columns parameter when calling the dataframe constructor (which accepts an array-like object) to specify the column name (or the column name will be set to 0 by default):

`
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df
`

### Instructions
1. Use the loop method from the last screen to calculate the mean mileage and mean price for each of the top brands, storing the results in a dictionary.
2. Convert both dictionaries to series objects, using the series constructor.
3. Create a dataframe from the first series object using the dataframe constructor.
4. Assign the other series as a new column in this dataframe.
5. Pretty print the dataframe, and write a paragraph analyzing the aggregate data.

In [ ]:
# Mean price series
mean_price_series = pd.Series(mean_brand_price)
print(mean_price_series)

In [ ]:
# Mean mileage
top_brand_mileage = {}

for b in top_brands:
    selected_rows = autos[autos["brand"] == b]
    mean = selected_rows["odometer_km"].mean()
    top_brand_mileage[b] = mean

print(top_brand_mileage)

In [27]:
# Mean mileage series
mean_mileage_series = (pd.Series(top_brand_mileage)
                      .sort_values(ascending=False)
                     )
print(mean_mileage_series)

bmw              132665.477606
mercedes_benz    130848.812603
opel             129864.317537
volkswagen       129101.023819
audi             128974.455297
renault          128356.979405
ford             125281.801008
dtype: float64


In [34]:
# Mean mileage dataframe
brand_dataframe = pd.DataFrame(mean_mileage_series, columns=['mean_mileage'])
print(brand_dataframe)

                mean_mileage
bmw            132665.477606
mercedes_benz  130848.812603
opel           129864.317537
volkswagen     129101.023819
audi           128974.455297
renault        128356.979405
ford           125281.801008


In [35]:
# Combining dataframe
brand_dataframe["mean_price"] = mean_brand_r

audi             9420.485099
mercedes_benz    8582.339055
bmw              8361.343440
volkswagen       5398.856038
ford             3448.153967
opel             2953.514986
renault          2420.492906
dtype: float64
